"Creating a simple plugin" guide (#12022)
ckeditor/ckeditor5: docs/framework/guides/simple-plugin-tutorial/abbreviation-plugin-level-2.md

---
category: simple-plugin
order: 25
---

# Abbreviation plugin tutorial - part 2

In this part of the tutorial we'll focus on creating a dialog box, which will get user's input.

We'll pick up where we left off in the first part, so make sure you {@link framework/guides/simple-plugin-tutorial/abbreviation-plugin-level-1 start there}, or grab our starter files for this part.

<info-box>
As we'll mostly work on the UI, we recommend reading up on our {@link framework/guides/architecture/ui-library UI library} before you start coding.
</info-box>

If you want to see the final product of this tutorial before you plunge in, check out the [demo](#demo).

## Creating a view

The most important part of the UI for this plugin is a dialog box with a form, which will get us user's input.

### Creating a form view template

Let's start by creating a view with a form. It will include two input fields (for the abbreviation and the title), and the 'submit' and 'cancel' buttons. We'll do it in a separate view. Create a new file `abbreviationview.js` in the `abbreviation/` directory.

> Our new `FormView` class will extend the {@link framework/guides/architecture/ui-library#views View} class, so start by importing it from the UI library.

Reviewer: Maybe we can add a link to the `FormView` class?

We'll also pass the editor's {@link module:utils/locale~Locale} instance to the constructor, so we can localize all our UI components with the help of the {@link module:utils/locale~Locale#t `t()` function}.

```js
// abbreviation/abbreviationview.js

import View from '@ckeditor/ckeditor5-ui';

export default class FormView extends View {
	constructor( locale ) {
		super( locale );
		const t = locale.t;

		this.setTemplate( {
			tag: 'form',
			attributes: {
				class: [ 'ck', 'ck-abbr-form' ],
				tabindex: '-1'
			}
		} );
	}

}
```
Notice that we added two classes. All UI elements of the editor need to have the `ck` class (unless you want to create your own UI and not use our library). We also created a new class for our form, which we'll use later on to style it.

### Creating input fields

As we have two similar input fields to create and we don't want to repeat ourselves, let's define a method `_createInput()`, which will produce them for us. It will accept the label of our input field.

> We'll use {@link module:ui/labeledinput/labeledfieldview~LabeledFieldView `LabeledFieldView`} class, and we'll pass it the `createLabeledInputText()` function as the second argument.

Reviewer: Maybe we can link to the `createLabeledInputText()` function and mention that this is a helper coming from the UI library that allows creating the input in an easier way?

```js
// abbreviation/abbreviationview.js

import {
    View,
    LabeledFieldView,               // ADDED
    createLabeledInputText          // ADDED
    } from '@ckeditor/ckeditor5-ui';

export default class FormView extends View {
	constructor( locale ) {
        // ...

        this.abbrInputView = this._createInput( t( 'Add abbreviation' ) );
		this.titleInputView = this._createInput( t( 'Add title' ) );
	}

    _createInput( label ) {
        const labeledInput = new LabeledFieldView( this.locale, createLabeledInputText );

        labeledInput.label = label;

        return labeledInput;
    }
}
```

### Creating form buttons

Now, we can add a `submit` and a `cancel` buttons to our form. You can start by importing `ButtonView` from our UI library and icons, which we'll use for labels.

Let's write a `_createButton` function, which will take three arguments - `label`, `icon` and `className`.

We'll then set the button attributes, using the properties we passed into the function, and adding a tooltip option.

> Last thing is to delegate `cancelButtonView#execute` to the FormView, so pressing it will fire off `FormView#cancel`.

Reviewer: Also, there is a new import - `{ icons }`. Can we add a sentence about it and add a link to available icons (probably somewhere in our repository)?

import { icons } from '@ckeditor/ckeditor5-core';   // ADDED

export default class FormView extends View {
	constructor( locale ) {
        // ...

        // Create the save and cancel buttons.
        this.saveButtonView = this._createButton(
            t( 'Save' ), icons.check, 'ck-button-save'
        );
		// Set the type to 'submit', which will trigger
		// the submit event on entire form when clicked.
		this.saveButtonView.type = 'submit';

		this.cancelButtonView = this._createButton(
            t( 'Cancel' ), icons.cancel, 'ck-button-cancel'
        );
		// Delegate ButtonView#execute to FormView#cancel.
		this.cancelButtonView.delegate( 'execute' ).to( this, 'cancel' );

	}

    _createInput( label ) {
        // ...
    }

    _createButton( label, icon, className ) {
		const button = new ButtonView( this.locale );

		button.set( {
			label,
			icon,
			tooltip: true,
			class: className
		} );

		return button;
	}
}
```
### Adding styles

You can now open `styles.css` and style the new UI elements. Let's add some padding to our form and use the [CSS grid layout](https://developer.mozilla.org/en-US/docs/Web/CSS/grid) to nicely display our four elements of the form.

> We'll use our set spacing variables to keep things uniform.

Reviewer: ```suggestion

We'll use our set spacing variables to keep things uniform.
```

```js
// abbreviation/abbreviationview.js

import {
    View,
    LabeledFieldView,
    createLabeledInputText,
    ButtonView,
    submitHandler                                   // ADDED
    } from '@ckeditor/ckeditor5-ui';
import { icons } from '@ckeditor/ckeditor5-core';

export default class FormView extends View {
	constructor( locale ) {

        // ...

        this.childViews = this.createCollection( [
			this.abbrInputView,
			this.titleInputView,
			this.saveButtonView,
			this.cancelButtonView
		] );

		this.setTemplate( {
			tag: 'form',
			attributes: {
                // ...
			},
			children: this.childViews               // ADDED
		} );
	}

    render() {
        super.render();

		// Submit the form when the user clicked the save button
		// or pressed enter in the input.
        submitHandler( {
            view: this
        } );
    }

    focus() {
        this.childViews.first.focus();
    }

    _createInput( label ) {
        // ...
    }

    _createButton( label, icon, className ) {
        // ...
    }
}
```

Our `FormView` is done! We can't see it just yet, so let's add it to our `AbbreviationUI` class.

## Adding the Contextual Balloon

> Our form needs to appear in a balloon, and we'll use the `ContextualBalloon` class from the CKEditor 5 UI library to make one.

Reviewer: Let's add a link to the `ContextualBalloon` class.

		const { t } = editor.locale;

		editor.ui.componentFactory.add( 'abbreviation', locale => {
			const button = new ButtonView( locale );

			button.label = t( 'Abbreviation' );
			button.tooltip = true;
			button.withText = true;

			this.listenTo( button, 'execute', () => {
				const title = 'What You See Is What You Get';
				const abbr = 'WYSIWYG';

				editor.model.change( writer => {
					editor.model.insertContent(
						writer.createText( abbr ),
						{ 'abbreviation': title }
					);

				} );
			} );

			return button;
		} );
	}
}
```
We'll need to change it quite a bit and add our `ContextualBalloon` and `FormView`. We'll need to make sure Contextual Balloon is required when making an instance of our `AbbreviationUI`, so let's start with that.

Let's write a basic `_createFormView()` function, just to create an instance of our `FormView` class (we'll expand it later).

We also need to create a function, which will give us the target position for our balloon from user's selection. We need to convert selected view range into DOM range. We can use {@link module:engine/view/domconverter~DomConverter#viewRangeToDom `viewRangeToDom()` method} to do so.

> Finally, let's add our balloon amd fpr, view to the `init()` method.

Reviewer: > balloon amd fpr

😄 ?

import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
import ButtonView from '@ckeditor/ckeditor5-ui/src/button/buttonview';
import ContextualBalloon from '@ckeditor/ckeditor5-ui';             // ADDED
import FormView from './abbreviationview';                          // ADDED

class AbbreviationUI extends Plugin {
	static get requires() {
		return [ ContextualBalloon ];
	}

	init() {
        const editor = this.editor;
		const { t } = editor.locale;

        // Create the balloon and the form view.
		this._balloon = this.editor.plugins.get( ContextualBalloon );
		this.formView = this._createFormView();

		editor.ui.componentFactory.add( 'abbreviation', locale => {
            // ...
		} );
	}

	_createFormView() {
		const editor = this.editor;
		const formView = new FormView( editor.locale );

		return formView;
	}

	_getBalloonPositionData() {
		const view = this.editor.editing.view;
		const viewDocument = view.document;
		let target = null;

        // Set a target position by converting view selection range to DOM
		target = () => view.domConverter.viewRangeToDom(
			viewDocument.selection.getFirstRange()
		);

		return {
			target
		};
	}
}
```

> We can now change what happens when we the user pushes the toolbar button. We'll replace inserting the hard-coded abbreviation.

Reviewer: ```suggestion
We can now change what happens when the user pushes the toolbar button. We'll replace inserting the hard-coded abbreviation.
```

> Let's write a function which will show our UI elements, adding the form view to our balloon and setting its position. Last thing is to focus the form view, so the user can immediately start typing in the first input field.

Reviewer: ```suggestion
Let's write a `_showUI()` method which will show our UI elements by adding the form view to our balloon and setting its position. The last thing is to focus the form view, so the user can immediately start typing in the first input field.
```

```js
// abbreviation/abbreviationui.js
// ...

class AbbreviationUI extends Plugin {
	static get requires() {
		return [ ContextualBalloon ];
	}

	init() {
        // ...
	}

	_createFormView() {
		const editor = this.editor;
		const formView = new FormView( editor.locale );

		this.listenTo( formView, 'submit', () => {
			const title = formView.titleInputView.fieldView.element.value;
			const abbr = formView.abbrInputView.fieldView.element.value;

			editor.model.change( writer => {
				editor.model.insertContent(
					writer.createText( abbr, { abbreviation: title } )
				);
			} );

		} );

		return formView;
	}

	_getBalloonPositionData() {
        // ...
	}

	_showUI() {
		// ...
	}
}
```
Our plugin is finally doing what it's supposed to. The last thing is to hide it from our editor when we don't need it.

## Hiding the form view

We'll need to hide the form view on three occasions:
* after the user submits the form;

> * when the user clicks the "Cancel" button; and

Reviewer: ```suggestion
* when the user clicks the "Cancel" button
```
